plugins/webp-uploads/tests/test-picture-element.php

```diff
@@ -176,10 +176,12 @@ public function test_img_sizes_is_equal_to_picture_source_sizes_for_picture_elem
 		$img_processor = new WP_HTML_Tag_Processor( $img_markup );
 		$this->assertTrue( $img_processor->next_tag( array( 'tag_name' => 'IMG' ) ), 'There should be an IMG tag.' );
 		$img_sizes = $img_processor->get_attribute( 'sizes' );
+		$img_src   = $img_processor->get_attribute( 'src' );
 
 		$picture_processor = new WP_HTML_Tag_Processor( $picture_markup );
 
 		$picture_processor->next_tag( array( 'tag_name' => 'IMG' ) );
+		$this->assertSame( $img_src, $picture_processor->get_attribute( 'src' ), 'Make sure IMG src is same.' );
```

Comment:
The `IMG` tag looks like this:

```html
<img width="1024" height="683" src="http://localhost:8889/wp-content/uploads/2024/07/leaves-950-1024x683.webp" class="wp-image-9" alt="Green Leaves" decoding="async" loading="lazy" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-950-1024x683.webp 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-950-300x200.webp 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-950-768x512.webp 768w, http://localhost:8889/wp-content/uploads/2024/07/leaves-950-jpg.webp 1080w" sizes="(max-width: 1024px) 100vw, 1024px" />
```

And below the `PICTURE` element looks like this:

```html
<picture class="wp-picture-9" style="display: contents;">
  <source type="image/webp" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-950-1024x683.webp 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-950-300x200.webp 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-950-768x512.webp 768w" sizes="(max-width: 1024px) 100vw, 1024px">
  <img width="1024" height="683" src="http://localhost:8889/wp-content/uploads/2024/07/leaves-950-1024x683.webp" class="wp-image-9" alt="Green Leaves" decoding="async" loading="lazy" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-950-1024x683.webp 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-950-300x200.webp 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-950-768x512.webp 768w, http://localhost:8889/wp-content/uploads/2024/07/leaves-950-jpg.webp 1080w" sizes="(max-width: 1024px) 100vw, 1024px" />
</picture>
```

A couple things seem off here:

1. The `IMG` has an identical `srcset` to the `PICTURE > SOURCE`, except the `IMG` has one additional entry in the `srcset`: `http://localhost:8889/wp-content/uploads/2024/07/leaves-950-jpg.webp 1080w`. Why?

2. Apparently JPEG fallbacks are not enabled, so this is generating a `PICTURE` for no reason, since both the `PICTURE > SOURCE` and the `PICTURE > IMG` are both serving WebP. If you look at `webp_uploads_is_picture_element_enabled()` you can see that it only returns true if both the the `perflab_generate_webp_and_jpeg` and `webp_uploads_use_picture_element` options are `true`, and so in `hooks.php` it only adds the `webp_uploads_wrap_image_in_picture` filter when they are both true. So to fix this in the test, I think the `opt_in_to_picture_element` helper should be changed to also do:

```php
update_option( 'perflab_generate_webp_and_jpeg', '1' );
```

When this is done, then the effect will be a unit test failure because when the picture element is not enabled, the `IMG` should be outputting the modern image format, whereas when the picture element is enabled then the `PICTURE > IMG` should _not_ be the modern image format but rather the JPEG fallback.

Reply:
@westonruter Thanks for the feedback.

> The IMG has an identical srcset to the PICTURE > SOURCE, except the IMG has one additional entry in the srcset: http://localhost:8889/wp-content/uploads/2024/07/leaves-950-jpg.webp 1080w. Why?

Per https://github.com/WordPress/performance/blob/trunk/plugins/webp-uploads/picture-element.php#L33-L42 the scrset is account only for sub sizes for the img that's why it get missed in PICTURE > SOURCE. I will open separate tracking issue for that. 


There is some juggling in unit tests it, also found some bug in code that replace original IMG to alternate mime type through `add_filter( 'the_content', 'webp_uploads_update_image_references', 13 )` filter.

For your information, I have update PR with the following:
- Removed the filter that add through function for unit test, instead use options that add the functionality.
```
public function opt_in_to_picture_element(): void {
	add_filter( 'wp_content_img_tag', 'webp_uploads_wrap_image_in_picture', 10, 3 );
}
```
- Remove the shared fixtures as it return WEBP image instead JPEG image. It because when image get uploaded it will update with the alternate mime type it was WEBP.
- Check IMG and Picture element IMG src should end with `.jpg` (as image is `leaves.jpg`)
- Remove `add_filter( 'the_content', 'webp_uploads_update_image_references', 13 )` filter that update IMG tag with alternate mime type.
- Check Picture element requirement condition in filter and bail early if not met.


**IMG markup:**
```
<img width="1024" height="683" src="http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-1024x683.jpg" class="wp-image-74" alt="Green Leaves" decoding="async" loading="lazy" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-1024x683.jpg 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-300x200.jpg 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-768x512.jpg 768w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087.jpg 1080w" sizes="(max-width: 1024px) 100vw, 1024px" />
```
**Picture markup:**
```
<picture class="wp-picture-74" style="display: contents;">
 <source type="image/webp" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-1024x683-jpg.webp 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-300x200-jpg.webp 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-768x512-jpg.webp 768w" sizes="(max-width: 1024px) 100vw, 1024px">
 <source type="image/jpeg" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-1024x683.jpg 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-300x200.jpg 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-768x512.jpg 768w" sizes="(max-width: 1024px) 100vw, 1024px">
 <img width="1024" height="683" src="http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-1024x683.jpg" class="wp-image-74" alt="Green Leaves" decoding="async" loading="lazy" srcset="http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-1024x683.jpg 1024w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-300x200.jpg 300w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087-768x512.jpg 768w, http://localhost:8889/wp-content/uploads/2024/07/leaves-1087.jpg 1080w" sizes="(max-width: 1024px) 100vw, 1024px" />
</picture>
```